### ISSUE TITLE
Single domain with RSA and ECDSA certs

### ISSUE THREAD

**Issue by madhanrajrm** [haproxy/haproxy#2392]
### Detailed Description of the Problem

promoting this issue form this discussion https://discourse.haproxy.org/t/single-domain-with-rsa-and-ecdsa-certs/8248

I use HAProxy 2.8, I have 2 certificates for RSA and ECDSA, it is set up as described in the docs . It used to work fine in 2.0 

certificate
[root@abctest~]# ls -ltr /usr/local/
-rwxr-xr-x. 1 certbase ccmbase 4119 Dec 13 09:00 HAProxy.pem.rsa
-rwxr-xr-x. 1 certbase ccmbase 1173 Dec 13 09:00 HAProxy.pem.ecdsa

here problem was when i connect to my port using below openssl commands it works fine .( it offers me back ECDSA cert).
openssl s_client -connect localhost:8443 -tls1_3

but all of my below executions it fails to connect.
openssl s_client -connect localhost:8443 -tls1_2 -cipher ECDHE-RSA-AES256-GCM-SHA384
openssl s_client -connect localhost:8443 -tls1_2 -cipher ECDHE-ECDSA-AES256-GCM-SHA384

fd[0x9] OpenSSL error[0x1417a0c1] tls_post_process_client_hello: no shared cipher
fd[0x9] OpenSSL error[0x1417a0c1] tls_post_process_client_hello: no shared cipher

openssl s_client -connect localhost:8443 -sigalgs RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512

fd[0x9] OpenSSL error[0x14201076] tls_choose_sigalg: no suitable signature algorithm

i have hidden the build number of openssl we use but it has all the vulnerability fixes.

### Expected Behavior

Ideally with hybrid configuration in TLS 1.2  since we are preferring client ciphers when i have  my client offers ECDSA ciphers  then server should send me back ECDSA certs and likewise for RSA . 

similarly for TLS 1.3 the same should be happening using sigalgs. 

### Steps to Reproduce the Behavior

1. Use latest haproxy 2.8.5
2. Try to configure hybrid configuration i.e., load both RSA and ECDSA certs on same port and ip.
3. in the bind line just give the base name of the certificates. 

### What is your configuration?

```haproxy
Here is my config

global
  maxconn 16000
  nbthread 1
  maxsslrate 500
  maxsslconn 18000
  tune.bufsize 32768
  tune.ssl.lifetime 3600
  log /dev/log local5

  stats socket ipv4@127.0.0.1:9999 level admin
  stats socket /var/run/haproxy.sock mode 666 level admin
  stats timeout 2m

  ssl-default-bind-ciphers ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:AES256-SHA:AES128-SHA:DHE-RSA-AES128-SHA
  ssl-default-bind-ciphersuites TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256
  ssl-default-bind-options ssl-min-ver TLSv1.0  ssl-max-ver TLSv1.3 no-tls-tickets

defaults
  log     global
  mode    http
  option  log-health-checks
  option  log-separate-errors
  option  dontlognull
  log-format '%tr %ci:%cp %fp/%fc/%Th/%Tq/%rc %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}HP'
  option  splice-auto
  option  socket-stats
  option  redispatch
  maxconn 5000
  timeout connect 3s
  timeout client 300s
  timeout server 600s
  timeout client-fin 1s
  timeout server-fin 1s



listen ssl


  bind 0.0.0.0:8443   tfo ssl crt /usr/local/HAProxy.pem  npn http/1.1 curves X25519:P-256:P-384:P-521 prefer-client-ciphers


  maxconn 30000
  rate-limit sessions 200
  option tcp-smart-connect
  option forwardfor
  http-request add-header X-Forwarded-Proto https if { ssl_fc }
  http-request set-header X-Forwarded-Port %[dst_port]
  default_backend idpserver

....
```

### Output of `haproxy -vv`

```plain
And my haproxy -vv

[root@abctext~]# ./haproxy -vv
HAProxy version 2.8.5-aaba8d0 2023/12/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.5.html
Running on: Linux 4.18.0-372.16.1.el8_6.x86_64 #1 SMP Wed Jul 13 03:56:16 EDT 2022 x86_64
Build options :
  TARGET  = custom
  CPU     = generic
  CC      = cc
  CFLAGS  = -m64 -march=x86-64 -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_EPOLL=1 USE_THREAD=1 USE_LINUX_SPLICE=1 USE_OPENSSL=1 USE_SLZ= USE_CPU_AFFINITY=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES -ACCEPT4 -BACKTRACE -CLOSEFROM +CPU_AFFINITY -CRYPT_H -DEVICEATLAS -DL -ENGINE +EPOLL -EVPORTS -GETADDRINFO -KQUEUE -LIBATOMIC -LIBCRYPT -LINUX_CAP +LINUX_SPLICE -LINUX_TPROXY -LUA -MATH -MEMORY_PROFILING -NETFILTER -NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL -PRCTL -PROCCTL -PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT -RT -SHM_OPEN -SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD -TFO +THREAD -THREAD_DUMP -TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 1.1.1t.***
Running on OpenSSL version : OpenSSL 1.1.1t.***
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): no
Built with gcc compiler version 8.5.0 20210514 (Red Hat 8.5.0-10)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by wlallemand** [comment]
Hello,

I think this happens because you didn't specify a servername in your command line, and there is no ECDSA/RSA selection on the default certificate, in version 2.0 of HAProxy, both certificates were in the same SSL_CTX so openSSL was selecting the right one itself, but that's not the case anymore. When there is no servername sent by the client the default certificate is always selected, this default certificate is always the first one, if you want to change this order you can write the configuration like this:

```bind 0.0.0.0:8443 ssl crt /usr/local/HAProxy.pem.rsa crt /usr/local/HAProxy.pem.ecdsa```

Could you reproduce with the "-servername" parameter?
```
openssl s_client -servername localhost -connect localhost:8443 -tls1_2 -cipher ECDHE-RSA-AES256-GCM-SHA384
openssl s_client -servername localhost -connect localhost:8443 -tls1_2 -cipher ECDHE-ECDSA-AES256-GCM-SHA384
```

**Comment by madhanrajrm** [comment]
> Hello,
> 
> I think this happens because you didn't specify a servername in your command line, and there is no ECDSA/RSA selection on the default certificate, in version 2.0 of HAProxy, both certificates were in the same SSL_CTX so openSSL was selecting the right one itself, but that's not the case anymore. When there is no servername sent by the client the default certificate is always selected, this default certificate is always the first one, if you want to change this order you can write the configuration like this:
> 
> `bind 0.0.0.0:8443 ssl crt /usr/local/HAProxy.pem.rsa crt /usr/local/HAProxy.pem.ecdsa`
> 
> Could you reproduce with the "-servername" parameter?
> 
> ```
> openssl s_client -servername localhost -connect localhost:8443 -tls1_2 -cipher ECDHE-RSA-AES256-GCM-SHA384
> openssl s_client -servername localhost -connect localhost:8443 -tls1_2 -cipher ECDHE-ECDSA-AES256-GCM-SHA384
> ```

Thanks @wlallemand  i tried this even with or without servername below behavior is consistent . i still dont feel sni is playing any role if i am studying the code properly. 

whatever you said seems to be happening here, it just picks only the default one (firstone) into ctx. If i load like below then just RSA algorithms/ciphers works because it just picks only first 
bind 0.0.0.0:8443 ssl crt /usr/local/HAProxy.pem.rsa crt /usr/local/HAProxy.pem.ecdsa

if i load the ecdsa certs first then it just picks only ECDSA algorithms/ciphers . 

bind 0.0.0.0:8443 ssl crt /usr/local/HAProxy.pem.ecdsa crt /usr/local/HAProxy.pem.rsa. 

So is there anyway i can load both the certificates into same context ? 


primarily i just want to ensure my ip is opening up an 8443 port and honors both the type of certificates.

**Comment by wlallemand** [comment]
@madhanrajrm is your certificate generated correctly? Do you have a CN and SNI in it? you could check with `openssl x509 -in file.pem -noout -text` .

> i still dont feel sni is playing any role if i am studying the code properly.

The whole code is based on SNI lookup in a tree in fact, so if your certificate doesn't have any domain in it, or if you client doesn't use the servername extension, it will fallback on the default.

**Comment by wlallemand** [comment]
> So is there anyway i can load both the certificates into same context ?

No, and you shouldn't care about this. If your need is to serve both ECDSA and RSA for domains which are not in your certificate then it's not supported currently. But that's a weird thing to want, since your configuration is using HTTP mode, HTTP clients are sending a servername, or you want to supports really old clients.

**Comment by wlallemand** [comment]
@madhanrajrm do you still have the issue? I didn't reproduce any problem with the TLS servername extension, neither with curl nor openssl. In my opinion there is no problem there, but do you have corrects CN and DNS fields in your certificates ? You could also test with the "strict-sni" keyword so it never serves a certificate which does not have the correct SNI.

**Comment by madhanrajrm** [comment]
@wlallemand  , Thanks yes our problem is pretty much peculiar  . 

we have pretty good old endpoints which doesn't offer SNI currently at the moment for HTTP connections. 

Currently based on my above config this is what is happening for us . 

1.  any HTTPS connection honored without SNI always defaults to first certificate it loads under default context . 
2. if SNI is offered  to our server if it matches with our  CN or SAN fields then i could see based on cipher / sigalgs preference corresponding type of cert gets offered back. 

I have an question though if i need to use haproxy as an front end to my non- HTTP server . is that possible ? in those scenarios  how does this tree implementations work ? 
Also can we move this issue to an enhancement request to change  the default certificates loading operations to  load both rsa and ecdsa certificates( if available) into default context so that openssl takes care of rest of the stuff ?

**Comment by wlallemand** [comment]
Okay that makes sense, since HAProxy 2.3 we don't put the whole certificate bundle (ECDSA+RSA+DSA) into a SSL_CTX, because it complexifies too much the architecture for dynamic updates over the CLI. The default_ctx can only be a SSL_CTX. I think we overlooked the default case when we made this change, however nobody complains because nobody tried to use ECDSA for clients that doesn't support SNI.

So indeed there is a change of behavior and this seems to be a problem with your usecase.

> I have an question though if i need to use haproxy as an front end to my non- HTTP server . is that possible ? in those scenarios how does this tree implementations work ?

TCP mode uses the same mechanism so it is also impacted. Basically every SNI is a key associated to a SSL_CTX in the tree. During the ClientHello a lookup is done using the servername field so we can chose the right SSL_CTX, if there are multiple entries it looks for the right ciphers and sigalgs match to chose between RSA or ECDSA. If there is no servername field, it fallback on the default SSL_CTX.

> Also can we move this issue to an enhancement request to change the default certificates loading operations to load both rsa and ecdsa certificates( if available) into default context so that openssl takes care of rest of the stuff ?

I agree that we could improve this. Technically I can't make everything in one SSL_CTX again, however I think I found a way to add multiple default in the tree with a special key so we can use the same algorithm when there is no servername to chose. I won't be able to backport this though.

**Comment by madhanrajrm** [comment]
cool , if it gets fixed it would be in 3.0 train ? and not anywhere in 2.8 or 2.9 versions ?

**Comment by wlallemand** [comment]
I don't think the changes will be small enough and without any impact so there is small chance it ends up backported in previous versions. I'll try to investigate a backportable solution, but so far I didn't find anything non-intrusive.

**Comment by wlallemand** [comment]
Hello, I made this branch https://github.com/haproxy/haproxy/tree/20230110-ssl-test-wildcard which allows a multi-certificate bundle to be the default, this is not merged in master yet, but you can test the feature. I don't think I will backport it previous versions unfortunately, this require to change the clienthello callback and that's a big change of behavior.

**Comment by madhanrajrm** [comment]
Thanks Let me check on this  and get back to you !!

**Comment by wlallemand** [comment]
I just pushed the code into master.

I also added a "default-crt" keyword on the bind line, allowing to have a clearer configuration of what is the default certificate.

**Comment by robert-scheck** [comment]
@wlallemand, as the 20230110-ssl-test-wildcard branch does not exist anymore, what's currently the best way to test this? Master branch? 3.0-dev branch? Or will this land in 3.0 at all?

**Comment by wlallemand** [comment]
@robert-scheck the 'default-crt' keyword was merged into master, it's already available in the master branch. Note that HAProxy 3.0 will probably be released this week.

**Comment by robert-scheck** [comment]
> @robert-scheck the 'default-crt' keyword was merged into master, it's already available in the master branch. Note that HAProxy 3.0 will probably be released this week.

Being not involved into HAProxy development, this leaves open for me whether this will be in HAProxy 3.0 or not (I do not keep track whether you've a 3.0 release branch and when the branch-off happened etc.).

**Comment by wlallemand** [comment]
@robert-scheck the master branch is always the branch that we use for the next major release. So 3.0 will be a fork of the master branch. So the default-crt keyword will be available in 3.0 indeed, as it was already available since 3.0-dev2.
